### PR TITLE
fix diff function

### PR DESCRIPTION
**Pull request by Thor-wl**
fix: https://github.com/volcano-sh/volcano/issues/1710
Signed-off-by: Thor-wl <13164644535@163.com>

### CODE REVIEW

File: pkg/scheduler/api/resource_info.go [hunk 1]
```diff
@@ -437,38 +437,43 @@ func (r *Resource) Equal(rr *Resource, defaultValue DimensionDefaultValue) bool
...
+			continue
 		}
-
-		if rQuant > rrQuant {
-			if increasedVal.ScalarResources == nil {
-				increasedVal.ScalarResources = map[v1.ResourceName]float64{}
-			}
-			increasedVal.ScalarResources[rName] += rQuant - rrQuant
+		if rQuant == -1 {
+			decreasedVal.ScalarResources[lName] = -1
```
Comment (huone1) [inline pkg/scheduler/api/resource_info.go:470]: I think it needn't to think about Infinity indiff function， just only set to zero ； if set to “-1” ， the logic using  this function also do the corresponding adjust

**Comment by huone1** [comment]
What's the reason?

File: pkg/scheduler/api/resource_info.go (hunk 1 above)
Comment (Thor-wl) [reply pkg/scheduler/api/resource_info.go:470]: As a tool function, it should be considered as general use for furture scenarios. For example, one of the comparsion object is queue.capability, dimensions not set should be considered as infinity.

**Comment by Thor-wl** [comment]
> What's the reason?

Diff function is not robust and lack of consideration about unequal dimensions between two objects.

File: pkg/scheduler/api/resource_info.go [hunk 2]
```diff
@@ -437,38 +437,43 @@ func (r *Resource) Equal(rr *Resource, defaultValue DimensionDefaultValue) bool
 }
 
 // Diff calculate the difference between two resource object
-func (r *Resource) Diff(rr *Resource) (*Resource, *Resource) {
+// Note: if `defaultValue` equals `Infinity`, the difference between two values will be `Infinity`, marked as -1
+func (r *Resource) Diff(rr *Resource, defaultValue DimensionDefaultValue) (*Resource, *Resource) {
+	leftRes := r.Clone()
```
Comment (william-wang) [inline pkg/scheduler/api/resource_info.go:442]: The Clone() function added recently causes 400ms+ latency in scheduling period while submit 1w pod.  We'd better think out a way without using Clone().

Comment (Thor-wl) [reply pkg/scheduler/api/resource_info.go:442]: > The Clone() function added recently causes 400ms+ latency in scheduling period while submit 1w pod. We'd better think out a way without using Clone().

OK. I think it's better to open another issue to optimize performance totally. As to this fix, the input parameter is a pointer and we should not make any modifications at the original object. So clone is needed.

File: pkg/scheduler/api/resource_info.go [hunk 3]
```diff
@@ -437,38 +437,43 @@ func (r *Resource) Equal(rr *Resource, defaultValue DimensionDefaultValue) bool
...
-		rrQuant, ok := rr.ScalarResources[rName]
-
-		if !ok {
-			rrQuant = 0
+	increasedVal.ScalarResources = make(map[v1.ResourceName]float64, 0)
+	decreasedVal.ScalarResources = make(map[v1.ResourceName]float64, 0)
+	for lName, lQuant := range leftRes.ScalarResources {
+		rQuant, _ := rightRes.ScalarResources[lName]
+		if lQuant == -1 {
+			increasedVal.ScalarResources[lName] = -1
```
Comment (jinzhejz) [inline pkg/scheduler/api/resource_info.go:466]: do we need to enhance `Resource.Sub()` and `Resource.Add()` when set `Resource.ScalarResources`  to -1?

File: pkg/scheduler/plugins/proportion/proportion.go [hunk 4]
```diff
@@ -178,7 +178,7 @@ func (pp *proportionPlugin) OnSessionOpen(ssn *framework.Session) {
 			klog.V(4).Infof("The attributes of queue <%s> in proportion: deserved <%v>, allocate <%v>, request <%v>, share <%0.2f>",
 				attr.name, attr.deserved, attr.allocated, attr.request, attr.share)
 
-			increased, decreased := attr.deserved.Diff(oldDeserved)
+			increased, decreased := attr.deserved.Diff(oldDeserved, api.Zero)
 			increasedDeserved.Add(increased)
 			decreasedDeserved.Add(decreased)
```
Comment (jinzhejz) [inline pkg/scheduler/plugins/proportion/proportion.go:183]: do we need to compare `increasedDeserved` and `increased` here?

File: pkg/scheduler/api/resource_info.go (hunk 3 above)
Comment (Thor-wl) [reply pkg/scheduler/api/resource_info.go:466]: > do we need to enhance `Resource.Sub()` and `Resource.Add()` when set `Resource.ScalarResources` to -1?

Actually yes, but that should be in PR focusing on the enhancement. That my be better.

File: pkg/scheduler/plugins/proportion/proportion.go (hunk 4 above)
Comment (Thor-wl) [reply pkg/scheduler/plugins/proportion/proportion.go:183]: > do we need to compare `increasedDeserved` and `increased` here?

I've not analysed the logic in proportion here. Just beacause of the input parameters changed, this call should be changed, too.

File: pkg/scheduler/api/resource_info.go (hunk 2 above)
Comment (william-wang) [reply pkg/scheduler/api/resource_info.go:442]: That's ok to me to open another issue to improve the performance.

**Review by william-wang — APPROVED** [review]

**Pull request merged**